### ISSUE TITLE
Add support for 4 byte UTF-8 characters and HTML entities over `&#xFFFF;` and `&#99999;`

### ISSUE THREAD

**Issue by stil** [libgd/libgd#185]
Currently, GD supports only UTF-8 characters of 1-3 bytes. Four byte characters are unsupported. HTML entities higher than  `&#xFFFF;` and `&#99999;` are also unsupported.

It makes impossible to draw emojis, which are Unicode characters higher than U+1F601. UTF-8 representation of emojis are at least 4 bytes long. I thought I could bypass this issue using HTML entities. However it seems, that they are bugged too.

Invalid code lies in `gdft.c` file, `gdTcl_UtfToUniChar()` function.

Sources:
* [Only 1-3 byte UTF-8 characters in `gdft.c`](https://github.com/libgd/libgd/blob/master/src/gdft.c#L340-L342)
* [Code responsible for HTML entities parsing in `gdft.c`](https://github.com/libgd/libgd/blob/master/src/gdft.c#L293-L314)
* [More information on the bug in PHP bugtracker](https://bugs.php.net/bug.php?id=51135)
* [Proposed patch from PHP bugtracker](https://bugs.php.net/patch-display.php?bug=51135&patch=bundled-libgd-non-bmp-entities&revision=1267806282&display=1)

Possibly related issue
---
Someone at PHP bugtracker suspects, that due to integer overflow, Unicode characters higher than 65536 might also be broken.
> The responses to this report seem to imply that ImageTtfText WIIL work correctly for characters in the range U+10000 - U+1869F (65536 - 99999). However, this does not seem to be the case. When I call it with `&#65537;` (U+10001), it produces an "missing" glyph, even using a font that contains the character (CODE2001.ttf).

**Comment by vapier** [comment]
can you attach a small file that illustrates the issue for us ?

**Comment by stil** [comment]
4 byte UTF-8 characters issue
----

Below you can see glyphs of Arial Unicode font.

![arial_unicode](https://i.imgur.com/w8GTBS1.png)

And here you see, how GD draws consecutive glyphs starting from `0xFFD0` and ending at `0x1001F`.

![drawing_glyphs](https://i.imgur.com/edden2l.png)

You can clearly see, that characters after 0xFFFF are incorrectly parsed.

Check UTF-8 byte representation of 0xFFFF and 0x10000 Unicode characters here:
* http://unicode-table.com/en/FFFF/
* http://unicode-table.com/en/10000/

**Comment by vapier** [comment]
i'm not saying you're wrong ... i'm asking for a small program that we can build/run to check the behavior on our side so that we can develop a fix & testcase.  that way we don't spend time recreating something you/someone else has already written, and we don't end up making changes for a different codepath w/out actually fixing the ones you're reporting.

**Comment by cmb69** [comment]
Adam Harvey already supplied a respective reproduce script including a respective font file and actual and expected images in the [respective PHP bug report](https://bugs.php.net/bug.php?id=51135), and the first halve of the necessary bugfixes. I have now analyzed the issue there further. TL;DR: we have to also select the appropriate charmap (a font can have several of them).

**Comment by pierrejoye** [comment]
Specifically for emoji, we need to implement the ft flag support. As far as I remember we have another issue for that.

**Comment by cmb69** [comment]
@pierrejoye Indeed there is issue #184, which would be a nice enhancement.

Wrt. this issue, it seems that in the long run we can simplify [finding the desired charmap](https://github.com/libgd/libgd/blob/gd-2.2.3/src/gdft.c#L1101-L1168) by just using `FT_Select_Charmap()` instead. For GD-2.2.x (and maybe GD-2.x.y) I would suggest to go with a minimal fix, namely to use `FT_SelectCharmap()` only if `encoding` is detected as `gdFTEX_Unicode`.

The biggest hurdle I'm currently facing is finding a font which we can use for the test case. Adam's test uses code2001.ttf which is apparently freeware, but it's lacking a license file.

**Comment by vapier** [comment]
you could always take an existing free font and hack it up with something like [FontForge](https://fontforge.github.io/) or [Birdfont](https://birdfont.org/).  we don't need the glyphs to be "correct", just unique so we can compare them against a reference image.

**Comment by 0biWanKenobi** [comment]
Hello, this seems to be still unfixed? Any ETA? Thanks

**Comment by cmb69** [comment]
> Any ETA?

No, sorry. A PR including a regression test would be very welcome!

**Comment by 0biWanKenobi** [comment]
>  A PR including a regression test would be very welcome! 

I'd love to help! Is there a way we can discuss what needs to be done, without polluting the thread? :)

**Comment by pentium10** [comment]
any update of this in 2019?

**Comment by sterlingpickens** [comment]
I've modified libgd and got it printing the entire 4 byte range now.
I'm still testing and am unsure about the html5 entities.
Also, I noticed xshow don't seem to account for the horiBearingX, is this intentional ie: meant to return the glyph origin ?

**Comment by sterlingpickens** [comment]
Here is a patch to test.
`diff -Naur ../libgd-2.3.2-orig/src/entities.h ./src/entities.h
--- ../libgd-2.3.2-orig/src/entities.h  2021-01-11 17:09:45.000000000 -0700
+++ ./src/entities.h    2021-04-17 06:44:00.325639616 -0700
@@ -269,7 +269,7 @@
                {"zwnj", 8204},
        };
 
-#define ENTITY_NAME_LENGTH_MAX 8
+#define ENTITY_NAME_LENGTH_MAX 9
 #define NR_OF_ENTITIES 252
 
 #ifdef __cplusplus
diff -Naur ../libgd-2.3.2-orig/src/gdft.c ./src/gdft.c
--- ../libgd-2.3.2-orig/src/gdft.c      2021-03-03 00:15:02.000000000 -0700
+++ ./src/gdft.c        2021-04-17 07:05:45.456881557 -0700
@@ -267,14 +267,43 @@
 static gdCache_head_t *fontCache;
 static FT_Library library;
 
-#define Tcl_UniChar int
-#define TCL_UTF_MAX 3
+#define Tcl_UniChar uint64_t
+#define TCL_UTF_MAX 4
+const unsigned char totalBytes[256] = {
+    1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
+    1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
+    1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
+    1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
+    1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
+    1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
+    2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,
+    3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,
+#if TCL_UTF_MAX > 3
+    4,4,4,4,4,4,4,4,
+#else
+    1,1,1,1,1,1,1,1,
+#endif
+
+#if TCL_UTF_MAX > 4
+    5,5,5,5,
+#else
+    1,1,1,1,
+#endif
+
+#if TCL_UTF_MAX > 5
+    6,6,6,6
+#else
+    1,1,1,1
+#endif
+};
+
 static int
 gdTcl_UtfToUniChar (const char *str, Tcl_UniChar * chPtr)
 /* str is the UTF8 next character pointer */
 /* chPtr is the int for the result */
 {
-       int byte;
+       unsigned char byte;
+       uint64_t b0, b1, b2, b3;
        char entity_name_buf[ENTITY_NAME_LENGTH_MAX+1];
        char *p;
        struct entities_s key, *res;
@@ -289,7 +318,7 @@
                if (byte == '#') {
                        byte = *((unsigned char *) (str + 2));
                        if (byte == 'x' || byte == 'X') {
-                               for (i = 3; i < 8; i++) {
+                               for (i = 3; i < 9; i++) {
                                        byte = *((unsigned char *) (str + i));
                                        if (byte >= 'A' && byte <= 'F')
                                                byte = byte - 'A' + 10;
@@ -302,7 +331,7 @@
                                        n = (n * 16) + byte;
                                }
                        } else {
-                               for (i = 2; i < 8; i++) {
+                               for (i = 2; i < 9; i++) {
                                        byte = *((unsigned char *) (str + i));
                                        if (byte >= '0' && byte <= '9')
                                                n = (n * 10) + (byte - '0');
@@ -400,9 +429,20 @@
 
                        *chPtr = (Tcl_UniChar) byte;
                        return 1;
+               } else if (byte < 0xF8) {
+                       b0=str[0];
+                       b1=str[1];
+                       b2=str[2];
+                       b3=str[3];
+                       if (  ((b1 & 0xC0) == 0x80) && ((b2 & 0xC0) == 0x80) && ((b3 & 0xC0) == 0x80)  ) {
+                               *chPtr = (Tcl_UniChar) (  ((b0 & 0x7) << 18) | ((b1 & 0x3F) << 12) | ((b2 & 0x3F) << 6) | (b3 & 0x3F));
+                               return 4;
+                       }
+                       *chPtr = (Tcl_UniChar)b0;
+                       return 1;
                }
-#if TCL_UTF_MAX > 3
-               else {
+#if TCL_UTF_MAX > 4
+                 else {
                        int ch, total, trail;
 
                        total = totalBytes[byte];
@@ -442,7 +482,7 @@
 } glyphInfo;
 
 static ssize_t
-textLayout(uint32_t *text, int len,
+textLayout(uint64_t *text, uint64_t len,
                FT_Face face, gdFTStringExtraPtr strex,
                glyphInfo **glyph_info)
 {
@@ -1105,12 +1145,12 @@
        FT_UInt glyph_index;
        double sin_a = sin (angle);
        double cos_a = cos (angle);
-       int  i, ch;
+       uint64_t  i, ch;
        font_t *font;
        fontkey_t fontkey;
-       const char *next;
+       char *next;
        char *tmpstr = 0;
-       uint32_t *text;
+       uint64_t *text;
        glyphInfo *info = NULL;
        ssize_t count;
        int render = (im && (im->trueColor || (fg <= 255 && fg >= -255)));
@@ -1300,7 +1340,7 @@
 #endif
        }
        if (encodingfound) {
-               FT_Set_Charmap(face, charmap);
+               FT_Select_Charmap(face, FT_ENCODING_UNICODE);
        } else {
                /* No character set found! */
                gdCacheDelete (tc_cache);
@@ -1315,17 +1355,17 @@
                        any2eucjp (tmpstr, string, BUFSIZ);
                        next = tmpstr;
                } else {
-                       next = string;
+                       next = (char *)string;
                }
 #ifndef JISX0208
        } else {
-               next = string;
+               next = (char *)string;
        }
 #endif
 
        oldpenf.x = oldpenf.y = 0; /* for postscript xshow operator */
        penf.x = penf.y = 0;    /* running position of non-rotated glyphs */
-       text = (uint32_t*) gdCalloc (sizeof (uint32_t), strlen(next));
+       text = (uint64_t *) gdCalloc (sizeof (uint64_t), (const char *)strlen(next));
        i = 0;
        while (*next) {
                int len;
@@ -1333,7 +1373,7 @@
                switch (encoding) {
                case gdFTEX_Unicode: {
                        /* use UTF-8 mapping from ASCII */
-                       len = gdTcl_UtfToUniChar (next, &ch);
+                       len = gdTcl_UtfToUniChar ((const char *)next, &ch);
                        /* EAM DEBUG */
                        /* TBB: get this exactly right: 2.1.3 *or better*, all possible cases. */
                        /* 2.0.24: David R. Morrison: use the more complete ifdef here. */
`

**Comment by sterlingpickens** [comment]
Here is a patch to test.
[Issue-185-01.diff](http://sterlingdesktops.com/pub/test/Issue-185-01.diff)

**Comment by sterlingpickens** [comment]
I refined the patch and extended it to include all html5 entities.
http://sterlingdesktops.com/pub/test/Issue-185-02.diff
I just need to do a complete test to verify that everything works.

**Comment by sterlingpickens** [comment]
http://sterlingdesktops.com/pub/test/Issue-185-04.diff
I verified that this one works with all the different html/utf input formats.

**Comment by vapier** [comment]
the [HTML spec](https://html.spec.whatwg.org/multipage/named-characters.html) links to a JSON database:
https://html.spec.whatwg.org/entities.json

seems like that'd be a lot easier to parse than an ad-hoc HTML parser ?  could even do it in Python pretty easily.

**Comment by sterlingpickens** [comment]
Who actually runs the generation script/program ?
I also worry that /usr/include/entities.h is not a good idea since libgd is the only thing using it(I could be wrong).
If this is the case, then we should compile a entities.h/.c only and not install it as a header.
The best solution might be to include a header that is already on most people's systems from a different project.
Also, that entities.json will more than double the array members because of case sensitivity and ";".

If we are going to be completely thorough then we should check that there are no conflicts with unicode as well. IE: & as the first byte matching some of these entities.

**Comment by vapier** [comment]
i agree we should stop installing entities.h.  i'll do that for the next major release.  this should only exist for our own gdft.c usage.

the generation script is for us devs.  we're committing the header file to the tree so we can ship it in releases.

**Comment by sterlingpickens** [comment]
gd_Entity_To_Unicode should be moved to entities.c as well, but then that'd complicate the gen script we end up using.

**Comment by vapier** [comment]
wrt entities.json, we can easily dedupe based on the trailing ; being omitted.  similarly, i think we can normalize the entity names to lowercase and call it a day.

**Comment by sterlingpickens** [comment]
"Entity names (the things that follow ampersands) are case-senstive, but many browsers will accept many of them entirely in uppercase or entirely in lowercase; a few must be cased in particular ways."

I think you're right, we have to go with entities.json.

4 different ways of specifying the same thing in this one example:
&Dot; &die; &DoubleDot; &uml;

My original solution was just taking the first, I hadn't considered this.

**Comment by sterlingpickens** [comment]
gdImageStringFTEx will have to be modified to handle two codepoints returned.
struct entities_s must have two ints to hold all these.

This will take some time. I will try not to flood this thread anymore untill I have something substantial.
I'm playing with json-c right now.

**Comment by vapier** [comment]
I would imagine python would be faster to do this in, so feel free to use that

**Comment by sterlingpickens** [comment]
I dunno about that, i'm already printing all the key/value pairs as strings in C.
It seems most the functions are just macros. The fastest might just be a bash script tbh.

I know perl better than I know python. perl regex would make quick work of it too.

**Comment by vapier** [comment]
we're in the process of removing perl from the codebase ;)

**Comment by sterlingpickens** [comment]
Ok, got a python generator working.
I need to modify gdft.c some more now

**Comment by sterlingpickens** [comment]
Everything is in place at this point. It just needs testing.

**Comment by pierrejoye** [comment]
I think we are good with the PR. We need tests, but I can give @sterlingpickens a hand for this. @sterlingpickens did an amazing job here and libgd will be better with this addition.

@vapier @cmb69 merge and I can help then, thought?

**Comment by cmb69** [comment]
Yeah, I suggest to merge PR #695, and to go from there.
